### PR TITLE
Fleet unify distributed training

### PR DESCRIPTION
**Pull request by seiriosPlus**

### CODE REVIEW

File: python/paddle/fluid/incubate/fleet/base/role_maker.py [hunk 1]
```diff
@@ -239,3 +246,32 @@ def _generate_role(self):
                 self.node_type_ = 1
             self.node_type_comm_ = self.comm_.Split(self.node_type_)
             self.role_is_generated_ = True
+
+
+class UserDefinedRoleMaker(RoleMakerBase):
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/base/role_maker.py:250]: I think you can add a local role maker that can be easily configured. For example, the ip is 127.0.0.0 by default and port id is random generated within a range.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                                export_for_deployment)
+
+    def save_persistables(self, executor, dirname, main_program=None):
+        io.save_persistables(executor, dirname, main_program, None)
+
+    def _get_worker_endpoints(self):
+        return self.worker_endpoints
+
+    def _get_current_id(self):
+        return self.current_id
```
Comment (sandyhouse) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py]: Should we expose the two methods to users (get_worker_endpoints and get_current_id).
To use paddle.fluid.compiler for distributed training, we have to set build_strategy.num_trainers and build_strategy.trainer_id. So, if we expose the above two methods to users, we can simplify the programming for users.

File: python/paddle/fluid/incubate/fleet/base/role_maker.py (hunk 1 above)
Comment (seiriosPlus) [reply python/paddle/fluid/incubate/fleet/base/role_maker.py:250]: yep, I will seriously consider.

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        self.server_endpoints = []
+        self.role = Role.WORKER
+        self.current_endpoint = None
+        self.current_id = 0
+        self.optimizer = None
+        self.role_maker_ = None
+
+    def is_first_worker(self):
+        """
+        Check whether the node is the first instance of worker.
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/base/fleet_base.py:63]: please add Examples here if you want to make this public to users.

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+        Returns:
+            bool: True if this is the first node of worker,
+                  False if not.
+        """
+        return self.is_worker() and self.current_id == 0
+
+    def worker_id(self):
+        """
+        Get current worker id.
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/base/fleet_base.py:73]: I think worker_id should be clarified here. In Collective mode, a worker id corresponds a gpu device card. In Parameter Server mode, worker id should be a pod that runs multi-thread training.

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+        Returns:
+            int: node id
+        """
+        return self.current_id
+
+    def worker_num(self):
+        """
+        Get current worker number.
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/base/fleet_base.py]: similar clarification with worker_id can be added in worker_num.

File: python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        """
+        this function saves the result from DistributedOptimizer.minimize()
+        """
+        self._opt_info = opt_info
+
+
+fleet = PSLib()
+
+
+class PSLibOptimizer(DistributedOptimizer):
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/parameter_server/pslib/__init__.py]: The name of PSLibOptimizer seems wired. Could you rename this with algorithmic description to this class?

**Review by guru4elephant — COMMENTED** [review]
generally I think more usage examples can be added. Although we do not want to release incubate to public currently, we have to do it in the future and users will need solid documentation of fleet.

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, mode):
+        """
+        Init Fleet with mode, mode must be an instance of Mode.
+
+        The `dirname` is used to specify the folder where persistable variables
+        are going to be saved. If you would like to save variables in separate
+        files, set `filename` None; if you would like to save all variables in a
+        single file, use `filename` to specify the file name.
```
Comment (sandyhouse) [inline python/paddle/fluid/incubate/fleet/base/fleet_base.py]: > The dirname is used to specify the folder where persistable variables
> are going to be saved. If you would like to save variables in separate
> files, set filename None; if you would like to save all variables in a ...

What does that mean? Or add these in wrong place?

**Comment by shanyi15** [comment]
have checked `distribute_transpiler.py` ,`dataset.py`, please preview `Optimizer` first.

**Review by sandyhouse — DISMISSED** [review]

File: python/paddle/fluid/optimizer.py [hunk 8]
```diff
@@ -38,11 +38,12 @@
 import copy
 
 __all__ = [
-    'SGD', 'Momentum', 'Adagrad', 'Adam', 'Adamax', 'DecayedAdagrad', 'Ftrl',
-    'SGDOptimizer', 'MomentumOptimizer', 'AdagradOptimizer', 'AdamOptimizer',
-    'AdamaxOptimizer', 'DecayedAdagradOptimizer', 'RMSPropOptimizer',
-    'FtrlOptimizer', 'Adadelta', 'ModelAverage', 'LarsMomentum',
-    'LarsMomentumOptimizer', 'DGCMomentumOptimizer'
+    'Optimizer', 'SGD', 'Momentum', 'Adagrad', 'Adam', 'Adamax',
```
Comment (chengduoZH) [inline python/paddle/fluid/optimizer.py]: Don't expose `Optimizer` here.

Comment (seiriosPlus) [reply python/paddle/fluid/optimizer.py]: done

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+    A user should pass a paddle.fluid.optimizer to DistributedOptimizer
+    minimize() function is implemented.
+    DistributedOptimizer is the starting point for a user who wants to
+    run distributed training. The optimized information will be stored in
+    Fleet() instance who holds the global information about current distributed
+    training.
+    """
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, optimizer, strategy=None):
```
Comment (chengduoZH) [inline python/paddle/fluid/incubate/fleet/base/fleet_base.py:256]: Please explain the args.

File: python/paddle/fluid/incubate/fleet/parameter_server/distributed_transpiler/__init__.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+from ...base.fleet_base import Fleet
+from ...base.fleet_base import Mode
+from ...base.fleet_base import DistributedOptimizer
+
+
+class DistributedTranspiler(Fleet):
+    def __init__(self):
+        """
+        A subclass for compatibility with fluid.transpiler.DistributeTranspiler.
+        """
```
Comment (chengduoZH) [inline python/paddle/fluid/incubate/fleet/parameter_server/distributed_transpiler/__init__.py]: line38~40 should be moved to between line36 and line 37.

File: paddle/fluid/API.spec [hunk 11]
```diff
@@ -26,7 +26,7 @@ paddle.fluid.DistributeTranspiler.get_pserver_program (ArgSpec(args=['self', 'en
...
 paddle.fluid.DistributeTranspiler.get_startup_program (ArgSpec(args=['self', 'endpoint', 'pserver_program', 'startup_program'], varargs=None, keywords=None, defaults=(None, None)), ('document', 'd796fc0c8d51503b556fcf6dc15c4f0c'))
 paddle.fluid.DistributeTranspiler.get_trainer_program (ArgSpec(args=['self', 'wait_port'], varargs=None, keywords=None, defaults=(True,)), ('document', '736330e31a7a54abccc0c7fd9119d9ff'))
-paddle.fluid.DistributeTranspiler.transpile (ArgSpec(args=['self', 'trainer_id', 'program', 'pservers', 'trainers', 'sync_mode', 'startup_program', 'current_endpoint'], varargs=None, keywords=None, defaults=(None, '127.0.0.1:6174', 1, True, None, '127.0.0.1:6174')), ('document', '06ce55338dfe96311ad1078235ab3bf4'))
+paddle.fluid.DistributeTranspiler.transpile (ArgSpec(args=['self', 'trainer_id', 'program', 'pservers', 'trainers', 'startup_program', 'current_endpoint'], varargs=None, keywords=None, defaults=(None, '127.0.0.1:6174', 1, None, '127.0.0.1:6174')), ('document', '951af0a910f9c264723da78ad555f3df'))
```
Comment (guru4elephant) [inline paddle/fluid/API.spec]: this change may affect previous examples I guess, could you check the API change for test_case or public examples?

File: python/paddle/fluid/incubate/fleet/base/fleet_base.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+
+from role_maker import RoleMakerBase, Role
+from role_maker import MPISymetricRoleMaker
+from role_maker import UserDefinedRoleMaker
+
+
+class Mode(Enum):
+    TRANSPILER = 1,
+    PSLIB = 2,
+    COLLECTIVE = 3
```
Comment (chengduoZH) [inline python/paddle/fluid/incubate/fleet/base/fleet_base.py:32]: Could you explain the difference between those modes?

Comment (seiriosPlus) [reply python/paddle/fluid/incubate/fleet/base/fleet_base.py:32]: Three distributed training mode.

Comment (seiriosPlus) [reply python/paddle/fluid/incubate/fleet/base/fleet_base.py:32]: I will add more information in `Mode `.

**Review by guru4elephant — APPROVED** [review]

**Pull request merged**